src/Microsoft.AspNetCore.Sockets.Client/Connection.cs

```diff
@@ -166,18 +173,36 @@ public Task StartAsync(ITransport transport, HttpClient httpClient)
         {
             try
             {
-                _logger.LogTrace("Beginning receive loop");
+                _logger.LogTrace("Beginning receive loop.");
 
                 while (await Input.WaitToReadAsync())
                 {
+                    if (_connectionState != ConnectionState.Connected)
```

Comment:
This prevents from trying to invoke receive events after the user called DisposeAsync. We also need to drain the channel otherwise Input.Completion task will never be completed. 

As for synchronization - we modify the state using InterlockedExchange and _connectionState is volatile. We change the state as the first thing if the user calls DisposeAsync or if the channel is closed by the transport.